qutebrowser/browser/downloads.py

```diff
@@ -894,110 +894,212 @@ def _on_data_changed(self, idx, *, webengine):
             qtutils.ensure_valid(end_index)
         self.dataChanged.emit(start_index, end_index)
 
-    def _raise_no_download(self, count):
-        """Raise an exception that the download doesn't exist.
+    def _select_downloads(self, count, all_, indexset, criterion):
+        """Return selected downloads specified by indexset that satisfy the
+        criterion.
 
         Args:
-            count: The index of the download
+            count: count of download to get the actions applied against.
+            all_: whether all downloads should get the actions applied against.
+            indexset: Specify the set of download indexes to delete. E.g.
+                '2,5,6-last,3', where 'last' is special keyword that denotes
+                that latest download index. If you specify numbers smaller than
+                1, or greater than 'last', then those numbers will be silently
+                ignored.
+            criterion: A function that returns True only when satisfied.
         """
-        if not count:
-            raise cmdexc.CommandError("There's no download!")
-        raise cmdexc.CommandError("There's no download {}!".format(count))
+
+        # support --all/-a
+        if all_ == True:
+            indexset = '1-last'
+
+        # support count
+        if indexset == None:
+            if count == 0:
+                count = len(self)
+            indexset = str(count)
+
+        # parse the indexset into numbers set
+        try:
+            indexset_p = utils.parse_numsettxt_into_list(indexset, len(self))
+        except ValueError:
+            raise cmdexc.CommandError(
+                    "Invalid index set '{}'!".format(indexset)
+                  )
+
+        # identify downloads that match the criterion, as well as those that
+        # don't exist
+        matchs = []
+        nomatchs = []
+        noexists = []
+        for i in indexset_p:
+            j = i-1 # j is the download item when counting from 0, i is the one
+                    # when counting from 1 - do we need this? maybe one day we
+                    # should count downloads, everywhere, from 0, and get away
+                    # with this?
+            try:
+                d = self[j]
+                if criterion(d):
+                    matchs.append(d)
+                else:
+                    nomatchs.append(i)
+            except IndexError:
+                noexists.append(i)
```

Comment:
No, it's fine to gather the downloads which didn't exit in a list - but why return that list and show the error outside of `_select_downloads` everywhere instead of doing it here and just returning matching downloads?